daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/svalue/Ordering.scala

```diff
@@ -0,0 +1,195 @@
+import com.digitalasset.daml.lf.language.Ast
+import com.digitalasset.daml.lf.speedy.SValue._
+import com.digitalasset.daml.lf.value.Value.AbsoluteContractId
+
+import scala.annotation.tailrec
+import scala.collection.JavaConverters._
+
+object Ordering extends scala.math.Ordering[SValue] {
+
+  private def zipAndPush[X, Y](
+      xs: Iterator[X],
+      ys: Iterator[Y],
+      stack: FrontStack[(X, Y)],
+  ): FrontStack[(X, Y)] =
+    (xs zip ys).to[ImmArray] ++: stack
+
+  private def compareIdentifier(name1: Ref.TypeConName, name2: Ref.TypeConName): Int = {
+    val c1 = name1.packageId compareTo name2.packageId
+    lazy val c2 = name1.qualifiedName.module compareTo name2.qualifiedName.module
+    def c3 = name1.qualifiedName.name compareTo name2.qualifiedName.name
+    if (c1 != 0) c1 else if (c2 != 0) c2 else c3
+  }
+
+  val builtinTypeIdx =
+    List(
+      Ast.BTUnit,
+      Ast.BTBool,
+      Ast.BTInt64,
+      Ast.BTText,
+      Ast.BTNumeric,
+      Ast.BTTimestamp,
+      Ast.BTDate,
+      Ast.BTParty,
+      Ast.BTContractId,
+      Ast.BTArrow,
+      Ast.BTOptional,
+      Ast.BTList,
+      Ast.BTTextMap,
+      Ast.BTGenMap,
+      Ast.BTAny,
+      Ast.BTTypeRep,
+      Ast.BTUpdate,
+      Ast.BTScenario
+    ).zipWithIndex.toMap
+
+  private def typeRank(typ: Ast.Type): Int =
+    typ match {
+      case Ast.TBuiltin(_) => 0
+      case Ast.TTyCon(_) => 1
+      case Ast.TNat(_) => 2
+      case Ast.TStruct(_) => 3
+      case Ast.TApp(_, _) => 4
+      case Ast.TVar(_) | Ast.TForall(_, _) | Ast.TSynApp(_, _) =>
+        throw new IllegalArgumentException(s"cannot compare types $typ")
+    }
+
+  @tailrec
+  // Any two ground types (types without variable nor quantifiers) can be compared.
+  private[this] def compareType(x: Int, stack0: FrontStack[(Ast.Type, Ast.Type)]): Int =
+    stack0 match {
+      case FrontStack() =>
+        x
+      case FrontStackCons(tuple, stack) =>
+        if (x != 0) x
+        else
+          tuple match {
+            case (Ast.TBuiltin(b1), Ast.TBuiltin(b2)) =>
+              compareType(builtinTypeIdx(b1) compareTo builtinTypeIdx(b2), stack)
+            case (Ast.TTyCon(con1), Ast.TTyCon(con2)) =>
+              compareType(compareIdentifier(con1, con2), stack)
+            case (Ast.TNat(n1), Ast.TNat(n2)) =>
+              compareType(n1 compareTo n2, stack)
+            case (Ast.TStruct(fields1), Ast.TStruct(fields2)) =>
+              compareType(
+                math.Ordering
+                  .Iterable[String]
+                  .compare(fields1.toSeq.map(_._1), fields2.toSeq.map(_._1)),
+                zipAndPush(fields1.iterator.map(_._2), fields2.iterator.map(_._2), stack)
+              )
+            case (Ast.TApp(t11, t12), Ast.TApp(t21, t22)) =>
+              compareType(0, (t11, t21) +: (t12, t22) +: stack)
+            case (t1, t2) =>
+              typeRank(t1) compareTo typeRank(t2)
```

Comment:
Maybe it would be nice to have a comment here explaining that this case can only occurs when t1 and t2 have different type rank, so we don't need to compare the rest of the stack. (And I would maybe add an assertion that `typeRank(t1) != typeRank(t2)`, given that changes to `typeRank(..)` would break this comparison function without any warning from the type system.)

Reply:
Good idea. added. Thanks

Reply:
Thanks! 👍